Introduce Custom Data support for ReceiveTlvs

shaavan opened the pull request:
**Part of** #1970

This PR introduces the ability to add custom data to `ReceiveTlvs`, enabling users to include this data in the blinded reply path they create for their counterparties. The custom data will be returned to the user once the blinded reply path is utilized, allowing users to attach and retrieve their custom data with the counterparty's Bolt12 message.

## Note on Follow-ups

While this PR establishes the core framework for custom data, additional related updates will be handled in follow-up PRs:

1. Updating the `ChannelManager` implementation for `create_blinded_path` to allow users to include custom data in reply paths created via `create_{offer, refund}_builder` and `pay_for_offer`.
2. Adding the ability to assess received custom data in the `message_received` function (feature dependent on #3525).

shaavan commented:
With the current version of PR, I am seeking concept/approach ACKs before I start working on the test suite.

A gentle ping, @jkczyz :)
Thanks!

valentinewallace commented on lightning/src/blinded_path/message.rs:
```diff
@@ -30,6 +32,8 @@ pub(crate) struct ReceiveTlvs {
 	/// the right context.
 	pub(crate) path_id: Option<[u8; 32]>,
+	/// Custom Tlvs
+	pub(crate) custom_tlvs: Vec<(u64, Vec<u8>)>,
```
> @valentinewallace is this one also LDK-specific in the way payment::ReceiveTlvs is?

Yep!

shaavan replied:
I was planning to handle the changes related to OnionMessage, and BlindedPayment seperately.
But it seems more logical now to handle them together.
Introduced custom_tlvs support for `payment::ReceiveTlvs` in c9d57f8345b4ec0a54a57718abcae59d11ccd343
Thanks!

valentinewallace replied:
> @valentinewallace is this one also LDK-specific in the way `payment::ReceiveTlvs` is?
> 
> In any case, we should really focus on including custom tlvs in `payment::ReceiveTlvs` (or, not sure if we need "generic custom TLVs" so much as "user provides one `Vec<u8>` that we put in one TLV).

@TheBlueMatt FYI, the BOLT 12 payment notifications scheme that Jeff wrote up includes blinded payment paths that aren't constructed by the receiver, so this may not be LDK specific anymore.

shaavan replied:
Updated the custom tlvs to be simple Vec<u8> in [pr2830.03](https://github.com/shaavan/rust-lightning/commits/pr2830.03)
Thanks for the suggestion!

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -22,6 +22,8 @@ pub(crate) struct ForwardTlvs {
 	/// introduction node with the blinded path.
 	pub(crate) next_blinding_override: Option<PublicKey>,
+	/// Custom Tlvs
+	pub(crate) custom_tlvs: Vec<(u64, Vec<u8>)>,
```
I agree with your logic here.
Addressed it in c9d57f8345b4ec0a54a57718abcae59d11ccd343

Thanks for catching it!

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -22,6 +22,8 @@ pub(crate) struct ForwardTlvs {
 	/// introduction node with the blinded path.
 	pub(crate) next_blinding_override: Option<PublicKey>,
+	/// Custom Tlvs
+	pub(crate) custom_tlvs: Vec<(u64, Vec<u8>)>,
```
Addressed in  c9d57f8345b4ec0a54a57718abcae59d11ccd343
Thanks!

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -30,6 +32,8 @@ pub(crate) struct ReceiveTlvs {
 	/// sending to. This is useful for receivers to check that said blinded path is being used in
 	/// the right context.
 	pub(crate) path_id: Option<[u8; 32]>,
+	/// Custom Tlvs
+	pub(crate) custom_tlvs: Vec<(u64, Vec<u8>)>,
```
Addressed in c9d57f8345b4ec0a54a57718abcae59d11ccd343
Thanks!

shaavan commented on lightning/src/onion_message/packet.rs:
```diff
@@ -282,12 +282,22 @@ pub(crate) enum ControlTlvs {
 impl Readable for ControlTlvs {
 	fn read<R: Read>(r: &mut R) -> Result<Self, DecodeError> {
-		_init_and_read_tlv_stream!(r, {
+		let mut custom_tlvs = Vec::new();
+
+		let tlv_len = BigSize::read(r)?;
+		let rd = FixedLengthReader::new(r, tlv_len.0);
+		_init_and_read_tlv_stream_with_custom_tlv_decode!(rd, {
 			(1, _padding, option),
 			(2, _short_channel_id, option),
 			(4, next_node_id, option),
 			(6, path_id, option),
 			(8, next_blinding_override, option),
+		}, |msg_type: u64, msg_reader: &mut FixedLengthReader<_>| -> Result<bool, DecodeError> {
+			if msg_type < 1 << 16 { return Ok(false) }
+			let mut value = Vec::new();
+			msg_reader.read_to_end(&mut value)?;
+			custom_tlvs.push((msg_type, value));
+			Ok(true)
```
Addressed  in c9d57f8345b4ec0a54a57718abcae59d11ccd343
Thanks!

shaavan commented:
@coderabbitai review

shaavan commented:
The are still some conceptual kinks to be ironed out before finishing up the custom TLV's implementation.
So Drafting up the PR for now! ♻️

shaavan commented:
**Updated from** [pr2380.02](https://github.com/shaavan/rust-lightning/commits/pr2830.02) to [pr2830.03](https://github.com/shaavan/rust-lightning/commits/pr2830.03) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2830.02..pr2830.03)):

### Changes:
- **Rebase** on main branch to resolve conflicts.
- **Introduce custom TLVs** as a simple Option<Vec<u8>>.
- **Expose an interface** to allow users to add custom TLVs to message::ReceiveTlvs.

### Note:
- The `Option<Vec<u8>>` field in `message::ReceiveTlvs` was chosen to support initializing Receive Payload with `None` fields, needed for a case in `fn packet_payloads_and_keys`.

I'm still refining the approach for custom TLVs within `payment::ReceiveTlvs`. Before moving forward with testing, I’d appreciate feedback on this initial approach. Thanks so much for your input!

shaavan commented on lightning/src/blinded_path/payment.rs:
```diff
@@ -141,18 +145,29 @@ impl<'a> Writeable for BlindedPaymentTlvsRef<'a> {
 
 		if let Some(short_channel_id) = scid {
 			if payment_secret.is_some() { return Err(DecodeError::InvalidValue) }
+			if !custom_tlvs.is_empty() { return Err(DecodeError::InvalidValue) }
```
Done! Corrected in [pr2830.03](https://github.com/shaavan/rust-lightning/commits/pr2830.03)
Thanks for the pointer!

shaavan commented on lightning/src/util/ser_macros.rs:
```diff
@@ -472,6 +472,7 @@ macro_rules! decode_tlv_stream {
 ///
 /// [`FixedLengthReader`]: crate::util::ser::FixedLengthReader
 /// [`DecodeError`]: crate::ln::msgs::DecodeError
+#[macro_export]
```
Corrected in [pr2830.03](https://github.com/shaavan/rust-lightning/commits/pr2830.03)
Thanks!

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -7939,6 +7939,7 @@ where
 				max_cltv_expiry,
 				htlc_minimum_msat: 1,
 			},
+			custom_tlvs: Vec::new(),
```
Exposed an interface to allow user pass custom tlvs in [pr2830.03](https://github.com/shaavan/rust-lightning/commits/pr2830.03)
Thanks!

shaavan commented:
**Updated from** [pr2380.02](https://github.com/shaavan/rust-lightning/commits/pr2830.02) to [pr2830.03](https://github.com/shaavan/rust-lightning/commits/pr2830.03) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2830.02..pr2830.03)):
Addressed @TheBlueMatt comments

### Changes:
- **Rebase** on main branch to resolve conflicts.
- **Introduce custom TLVs** as a simple Option<Vec<u8>>.
- **Expose an interface** to allow users to add custom TLVs to message::ReceiveTlvs.

### Note:
- The `Option<Vec<u8>>` field in `message::ReceiveTlvs` was chosen to support initializing Receive Payload with `None` fields, needed for a case in `fn packet_payloads_and_keys`.

I'm still refining the approach for custom TLVs within `payment::ReceiveTlvs`. Before moving forward with testing, I’d appreciate feedback on this initial approach. Thanks so much for your input!

shaavan commented:
**Updated from** [pr2380.03](https://github.com/shaavan/rust-lightning/commits/pr2830.03) to [pr2830.04](https://github.com/shaavan/rust-lightning/commits/pr2830.04) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2830.03..pr2830.04)):

Changes:
1. Introduce the sender_custom_tlvs & user_custom_tlvs distinction.
2. Introduce custom_tlvs in `payment::ReceiveTlvs`

shaavan commented:
**Updated from** [pr2380.04](https://github.com/shaavan/rust-lightning/commits/pr2830.04) to [pr2830.05](https://github.com/shaavan/rust-lightning/commits/pr2830.05) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2830.04..pr2830.05)):

Changes:
1. Finalise & corrected the approach.
2. Introduced test.

shaavan commented:
**Updated from** [pr2380.04](https://github.com/shaavan/rust-lightning/commits/pr2830.04) to [pr2830.05](https://github.com/shaavan/rust-lightning/commits/pr2830.05) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2830.04..pr2830.05)):

Changes:
1. Rebase on main.

jkczyz commented on lightning/src/blinded_path/message.rs:
```diff
@@ -241,7 +241,10 @@ pub(crate) struct ReceiveTlvs {
+	pub context: Option<MessageContext>,
+
+	/// Custom Tlvs. A user can use this to send custom tlvs information back to themself.
+	pub custom_tlvs: Option<Vec<u8>>,
```
Seems like this is just custom data rather than custom TLVs. Nothing is enforcing that this data is encoded as a TLV stream.

@TheBlueMatt Not sure if you have any thoughts on how if we should make this a `Vec<(u64, Vec<u8>)>` instead? And, if so, if we should encode it inside type `65539` or place them directly in `ReceiveTlvs`, restricting TLVs >= 2^16 and all the checks that that entails.

shaavan replied:
For now, I've updated the documentation to include a clear note on Forward Compatibility.
Let me know if we prefer moving to the `Vec<(u64, Vec<u8>)>` approach—I’d would love to integrate that!

jkczyz commented on lightning/src/ln/msgs.rs:
```diff
@@ -2739,13 +2743,18 @@ impl<'a> Writeable for OutboundOnionPayload<'a> {
 				// We need to update [`ln::outbound_payment::RecipientOnionFields::with_custom_tlvs`]
 				// to reject any reserved types in the experimental range if new ones are ever
 				// standardized.
+				let user_custom_tlv = (!user_custom_tlvs.is_empty()).then(|| (65537, user_custom_tlvs.to_vec()));
```
Why do we need to allocate a new `Vec` here?

shaavan replied:
After discussion offline, we concluded that a vec allocation is necessary here.

shaavan commented:
**Updated from** [pr2380.06](https://github.com/shaavan/rust-lightning/commits/pr2830.06) to [pr2830.07](https://github.com/shaavan/rust-lightning/commits/pr2830.07) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2830.06..pr2830.07)):

Changes:
1. Rebase on main to resolve merge conflicts.

shaavan commented:
**Updated from** [pr2380.07](https://github.com/shaavan/rust-lightning/commits/pr2830.07) to [pr2830.08](https://github.com/shaavan/rust-lightning/commits/pr2830.08) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2830.07..pr2830.08)):
Addressed @jkczyz comments

Changes:
1. Updated documentation.
2. Fix CI.

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -241,7 +241,10 @@ pub(crate) struct ReceiveTlvs {
-	pub context: Option<MessageContext>
+	pub context: Option<MessageContext>,
+
+	/// Custom Tlvs. A user can use this to send custom tlvs information back to themself.
```
Updated the documentation to be more precise and clearer in [pr2830.08](https://github.com/shaavan/rust-lightning/commits/pr2830.08)
Thanks for the pointer!

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -187,7 +187,9 @@ pub enum PendingHTLCRouting {
 		/// [`RecipientOnionFields::custom_tlvs`].
-		custom_tlvs: Vec<(u64, Vec<u8>)>,
+		sender_custom_tlvs: Vec<(u64, Vec<u8>)>,
+		/// Custom TLVs which were set by us through the reply path
```
I've updated the docs to clarify how `user_custom_tlvs` will be used in [pr2830.08](https://github.com/shaavan/rust-lightning/commits/pr2830.08). Let me know if they look good! Thanks!

shaavan commented:
**Updated from** [pr2380.08](https://github.com/shaavan/rust-lightning/commits/pr2830.08) to [pr2830.09](https://github.com/shaavan/rust-lightning/commits/pr2830.09) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2830.08..pr2830.09)):
Addressed @TheBlueMatt comments.

Changes:
1. Rename custom_tlvs -> custom_data.
2. Update documentation.

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -236,12 +236,24 @@ pub(crate) struct ForwardTlvs {
+	/// ## Note on Forward Compatibility:
+	/// Users can encode any kind of data into the `Vec<u8>`. However, they must ensure that the
+	/// data is structured in a forward-compatible manner to prevent data loss or misinterpretation
+	/// during future upgrades.
```
Updated the docs in [pr2830.09](https://github.com/shaavan/rust-lightning/commits/pr2830.09) for both message ReceiveTlvs, and payment ReceiveTlvs.
Thanks a lot for the pointer!

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -236,12 +236,24 @@ pub(crate) struct ForwardTlvs {
-	pub context: Option<MessageContext>
+	pub context: Option<MessageContext>,
+
+	/// Custom TLVs set by the user. If `custom_tlvs` is `Some`, it will be returned when the
```
Make sense! :)
Updated the field name from custom_tlvs -> custom_data in [pr2830.09](https://github.com/shaavan/rust-lightning/commits/pr2830.09)
Thanks for the suggestion!

shaavan commented:
**Updated from** [pr2380.09](https://github.com/shaavan/rust-lightning/commits/pr2830.09) to [pr2830.10](https://github.com/shaavan/rust-lightning/commits/pr2830.10) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2830.09..pr2830.10)):

Changes:
1. Rebase on main, to resolve merge conflicts.
2. Fix CI.

jkczyz commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -1633,7 +1644,7 @@ where
 							}
 						};
-						let response_instructions = self.offers_handler.handle_message(msg, context, responder);
+						let response_instructions = self.offers_handler.handle_message(msg, context, custom_data, responder);
```
We need to give the data to all handlers, not just the offers handler.

shaavan replied:
Done! [pr2830.11](https://github.com/shaavan/rust-lightning/commits/pr2830.11)
Thanks for the pointer!

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10452,8 +10452,13 @@ where
 
+		let receive_tlvs = message::ReceiveTlvs {
+			context: Some(context),
+			custom_data: None,
```
We'll want to let users set this, too, right? Probably best to do in a follow-up after #3412 is landed, but we will eventually need ways for them to add it to offer/refund message paths and reply paths (e.g., when paying for an offer). Let's update the PR description to mention what is done in this PR and what will be done in a follow-up.

shaavan replied:
Makes sense!
I’ve added a "TODO" comment in the code in [pr2830.11](https://github.com/shaavan/rust-lightning/commits/pr2830.11) and updated the PR description accordingly.

Thanks so much for pointer!

shaavan commented:
**Updated from** [pr2380.10](https://github.com/shaavan/rust-lightning/commits/pr2830.10) to [pr2830.11](https://github.com/shaavan/rust-lightning/commits/pr2830.11) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2830.10..pr2830.11)):
Addressed @jkczyz comments

**Changes:**
1. Restructure & Split commits to be more atomic.
2. Introduce custom_data for all the handlers.

shaavan commented:
**Updated from** [pr2380.11](https://github.com/shaavan/rust-lightning/commits/pr2830.11) to [pr2830.12](https://github.com/shaavan/rust-lightning/commits/pr2830.12) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2830.11..pr2830.12)):

Changes:
1. Rebase on main to resolve merge conflicts.
2. Change tlv number of every `user_custom_data` to 65541, to keep it consistent through the codebase.

shaavan commented:
**Updated from** [pr2380.11](https://github.com/shaavan/rust-lightning/commits/pr2830.11) to [pr2830.12](https://github.com/shaavan/rust-lightning/commits/pr2830.12) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2830.11..pr2830.12)):

Changes:
1. Rebase on main to resolve merge conflicts.
2. Change tlv number of every `user_custom_data` to 65541, to keep it consistent through the codebase.

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -236,12 +236,26 @@ pub(crate) struct ForwardTlvs {
-/// Similar to [`ForwardTlvs`], but these TLVs are for the final node.
-pub(crate) struct ReceiveTlvs {
+#[derive(Clone)]
+/// Similar to Forward Tlvs, but these TLVs are for the final node.
```
Addressed in [pr2830.11](https://github.com/shaavan/rust-lightning/commits/pr2830.11)
Thanks for the pointer!

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -236,12 +236,26 @@ pub(crate) struct ForwardTlvs {
-	pub context: Option<MessageContext>
+	pub context: Option<MessageContext>,
+
+	/// Custom data set by the user. If `custom_data` is `Some`, it will be returned when the
```
Done! [pr2830.11](https://github.com/shaavan/rust-lightning/commits/pr2830.11)
Thanks!

shaavan commented on lightning/src/blinded_path/message.rs:
```diff
@@ -236,12 +236,26 @@ pub(crate) struct ForwardTlvs {
+	/// Custom data set by the user. If `custom_data` is `Some`, it will be returned when the
+	/// blinded path is used.
+	///
+	/// This field allows encoding custom data intended to be retrieved when the blinded path is used.
```
Done! [pr2830.11](https://github.com/shaavan/rust-lightning/commits/pr2830.11)
Thanks!

shaavan commented on lightning/src/onion_message/messenger.rs:
```diff
@@ -402,7 +405,7 @@ pub struct ResponseInstruction {
 	/// [`OnionMessenger::send_onion_message_internal`] somewhat.
 	destination: Destination,
-	context: Option<MessageContext>,
+	reply_data: Option<(MessageContext, Vec<u8>)>,
```
Makes sense!
Addressed in [pr2830.11](https://github.com/shaavan/rust-lightning/commits/pr2830.11)
Thanks a lot for the pointer!

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -223,8 +229,13 @@ pub enum PendingHTLCRouting {
+		/// Custom data set by the receiver in the blinded path used to reach them.
+		///
+		/// For HTLCs received by LDK, these will ultimately bubble back up as
+		/// [`RecipientOnionFields::user_custom_data`].
```
Done! [pr2830.11](https://github.com/shaavan/rust-lightning/commits/pr2830.11)
Thanks!

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -12095,7 +12105,7 @@ where
 						let hmac = payment_hash.hmac_for_offer_payment(nonce, expanded_key);
 						let context = MessageContext::Offers(OffersContext::InboundPayment { payment_hash, nonce, hmac });
-						Some((OffersMessage::Invoice(invoice), responder.respond_with_reply_path(context)))
+						Some((OffersMessage::Invoice(invoice), responder.respond_with_reply_path(context, custom_data)))
```
Got it, makes sense! Fixed in [pr2830.11](https://github.com/shaavan/rust-lightning/commits/pr2830.11).
Thanks so much for the suggestion!

We can use the interface introduced in #3525 to assess the custom data along with the received Bolt12 message. I’ll take some time to think this through while wrapping up this PR.

shaavan commented:
**Updated from** [pr2380.12](https://github.com/shaavan/rust-lightning/commits/pr2830.12) to [pr2830.13](https://github.com/shaavan/rust-lightning/commits/pr2830.13) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2830.12..pr2830.13)):
Addressed @TheBlueMatt comments

Changes:
1. Remove redundant todos.
2. Introduce `custom_data` in only appropriate cases, aka. `InboundOnionPayload::BlindedReceive`.

shaavan commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10502,8 +10514,14 @@ where
 			.map(|(node_id, _)| *node_id)
 			.collect::<Vec<_>>();
 
+		// todo: Introduce ability to add custom_data here.
```
Done [pr2830.13](https://github.com/shaavan/rust-lightning/commits/pr2830.13)
Thanks!

shaavan commented on lightning/src/ln/msgs.rs:
```diff
@@ -1770,7 +1770,8 @@ mod fuzzy_internal_msgs {
 			keysend_preimage: Option<PaymentPreimage>,
-			custom_tlvs: Vec<(u64, Vec<u8>)>,
+			sender_custom_tlvs: Vec<(u64, Vec<u8>)>,
+			user_custom_data: Vec<u8>,
```
Removed the added `custom_data` in [pr2830.13](https://github.com/shaavan/rust-lightning/commits/pr2830.13)
Thanks a lot for the pointer!

shaavan commented on lightning/src/ln/msgs.rs:
```diff
@@ -1828,7 +1831,8 @@ mod fuzzy_internal_msgs {
 			keysend_preimage: Option<PaymentPreimage>,
-			custom_tlvs: &'a Vec<(u64, Vec<u8>)>,
+			sender_custom_tlvs: &'a Vec<(u64, Vec<u8>)>,
+			user_custom_data: &'a Vec<u8>,
```
Removed the added `custom_data` in [pr2830.13](https://github.com/shaavan/rust-lightning/commits/pr2830.13)
Thanks a lot for the pointer!

shaavan commented on lightning/src/ln/msgs.rs:
```diff
@@ -1813,7 +1815,8 @@ mod fuzzy_internal_msgs {
 			keysend_preimage: Option<PaymentPreimage>,
-			custom_tlvs: &'a Vec<(u64, Vec<u8>)>,
+			sender_custom_tlvs: &'a Vec<(u64, Vec<u8>)>,
+			user_custom_data: &'a Vec<u8>,
```
Removed the added `custom_data` in [pr2830.13](https://github.com/shaavan/rust-lightning/commits/pr2830.13)
Thanks a lot for the pointer!

shaavan commented:
**Updated from** [pr2380.13](https://github.com/shaavan/rust-lightning/commits/pr2830.13) to [pr2830.14](https://github.com/shaavan/rust-lightning/commits/pr2830.14) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2830.13..pr2830.14)):

Changes:
1. Rebase on main to resolve merge conflicts.

shaavan commented:
**Updated from** [pr2380.14](https://github.com/shaavan/rust-lightning/commits/pr2830.14) to [pr2830.15](https://github.com/shaavan/rust-lightning/commits/pr2830.15) ([diff](https://github.com/shaavan/rust-lightning/compare/pr2830.14..pr2830.15)):

Changes:
1. Fix CI

jkczyz commented on lightning/src/ln/outbound_payment.rs:
```diff
@@ -650,22 +651,25 @@ pub struct RecipientOnionFields {
+	/// See [`Self::sender_custom_tlvs`] for more info.
+	pub(super) sender_custom_tlvs: Vec<(u64, Vec<u8>)>,
+	/// See [`Self::user_custom_data`] for more info.
+	pub(super) user_custom_data: Option<Vec<u8>>
```
The PR has ways to include custom data in blinded path as well as a blinded payment path. The former is given to the user via a message handler argument.

Here, for the latter, we were thinking the data would go in the blinded payment path of the invoice, rather than the offer. And the forthcoming `OffersMessageFlow` would allow the user to examine the invoice request and add the data when generating the invoice. Including it in the offer could bloat its size.

> then when we generate an invoice request we copy the data into the blinded payment path then when we receive payment we bubble it up to the user

I think you meant "invoice" not "invoice request" here? I guess the question is how to bubble it up to the user. Using `RecipientOnionFields` (via `PaymentClaimable`) or something else?

shaavan commented:
Closing this out for now, since it's been deprioritised and hasn't progressed in quite a while. Based on the last update, it was clear some re-alignment on goals was needed, but given current priorities, I think it's best to revisit this only if it becomes relevant again.
